### ISSUE TITLE
Generic type changes depending if its parameter type is wrapped or not.  

### ISSUE THREAD

**Issue by miguel-leon** [microsoft/TypeScript#56675]
### 🔎 Search Terms

generic type bug

### 🕗 Version & Regression Information

latest version as of today.

### ⏯ Playground Link

https://www.typescriptlang.org/play?#code/C4TwDgpgBAkgZjAJhAdsAlgYwIYBsA8AGgDRQCapAglALxTABOArhKQEK1Rx4DOEAfLQBQUUVAAU+ACr9xASlqCpUCAA9gqRDyiEoAfigBGKAC4oAJgVqNKLSLGSZ8xVGXXN2svqOmLCg9RmbADcQmGgkLBwAPIM+AAKpACKpFLM0HSMLKQAYrwZXPmCdPBIqBg4BImFuHykpchoWHj4KTV1UHm1rK7p-Kl9oUIR0AD6xiUxcVk9M-zBUAD0i-Tpw+Bj5pzwsfgzpNzd80srM+uRowDM21P4hx1zC8urLOdjACw3u-c9P8fPPzCb1gPAAyox0CgAObSYpRMpNSrSUg8CHQ+bhDZQUYAVm2YLRMJQTAAtgAjCAMDFAkbYqQQVETKLffL1AkMSFE0kUqn-FY-AB0UEwAHsGAwIJhgABCIHAnYMADq6GAAAt4OCOdDYV84j82ZrObDqbTRvTUVtJrFlWqNYT8MTyZS+S8IAYyUwobKgA

### 💻 Code

```ts
type IfIdentical<X, Y, A = true, B = false> =
    (<T>() => T extends X ? 1 : 2) extends
    (<T>() => T extends Y ? 1 : 2) ? A : B;


type IfOr<P, Q, True = true, False = false> = IfIdentical<P, false, IfIdentical<Q, false, False, True>, True>;

type _1 = IfOr<true, true>; // true
type _2 = IfOr<true, false>; // true
type _3 = IfOr<false, true>; // true
type _4 = IfOr<false, false>; // false


type IsString<T> = IfIdentical<T, string>;

type _5 = IsString<number>;



type _Test1 = IfOr<false, IsString<number>>; // false. correct!



type IfOrWithIfString<T> = IfOr<false, IsString<T>>;

type _Test2 = IfOrWithIfString<number>; // true? bug!
```

### 🙁 Actual behavior

types `_Test1` and `_Test2` are different even though they are calculated the same way.

### 🙂 Expected behavior

types `_Test1` and `_Test2` should be the same.

### Additional information about the issue

Not sure if it's the best minimal reproduction, because it's difficult to pinpoint the actual problem.
The example is based on the a type suggested by Matt McCutchen [here](https://github.com/Microsoft/TypeScript/issues/27024#issuecomment-421529650), but I can't tell if the issue is unique to that type or if it happens in other scenarios.

Based on that `IsEqual` type I created a "if or" type `IfOr<P, Q>` and a "is string" type `IsString<T>`.

The difference between `_Test1` and `_Test2` is that the first passes `IsString<number>` result (`false`) directly to `IfOr`, while the second passes `number` to an intermediary type, sort of wrapping the `IsString`.

**Comment by RyanCavanaugh** [comment]
`IsIdentical` and/or `IsOr` is written incorrectly, resulting in a difference between argument-base and structure-based inference, same as might occur if you wrote `interface DoesNotUseT<T> { x: string }` and compared inference with `DoesNotUseT<number> - DoesNotUseT<U>` to `{ x: string } - DoesNotUseT<U>`. In both cases, the exact order of when things are instantiated can cause the result to change. The fix is to write `IsIdentical` correctly.

The divergence in expectation actually starts one line earlier:
```ts
type IfOrWithIsString<T> = IfOr<false, IsString<T>>;
//   ^?
//     type IfOrWithIsString = true, wat?

// Which branch of IfOr is being hit? Let's label them
type IfOrSimpler1<P, Q> = IfIdentical<P, false, IfIdentical<Q, false, "Both False", "Q is not false">, "P is not false">;
type IfOrSimpler1WithIfString<T> = IfOrSimpler1<false, IsString<T>>;
//   ^?
//   "Q is not false"
// It's hitting IsIdentical<IsString<T>, false>. Why?

// Simplify IfOr to just the P case
type IfOrSimpler2<Q> =  IfIdentical<Q, false, "Both False", "Q is not false">;
type IfOrSimpler2WithIsString<T> = IfOrSimpler2<IsString<T>>;
//   ^?
//   ^? "Q is not false"
// IfIdentical<IsString<T>, false> hits the 'false' condition

type IfIdenticalSimpler<X, Y> =
    (<T>() => T extends X ? 1 : 2) extends
    (<T>() => T extends Y ? 1 : 2) ? "Seems to extend" : "Seems not to";
type IfSimple3<Q> =  IfIdenticalSimpler<Q, false>;
type IfSimple3WithIsString<T> = IfSimple3<IsString<T>>;
//   ^?
//   ^? "Seems not to"
```

You could instead write this, which works more reliably:
```ts
type IfOr<P, Q, True = true, False = false> = IfIdentical<P, true, True, IfIdentical<Q, true, True, False>>;
```
or use a more usual definition of `IsIdentical`:
```ts
type IfIdentical<X, Y, A = true, B = false> = X extends Y ? Y extends X ? A : B : B;
```

TL;DR you can't take falsy results of conditional type evaluations as definitive. Restrictive instantiations can and do occur; for *most* conditional types, they should be of the form `A extends B ? X : X | Y` to produce correct results.

**Comment by RyanCavanaugh** [comment]
cc @weswigham please check my work on the above 😅

**Comment by sliminality** [comment]
I'm not sure I follow this:

> you can't take falsy results of conditional type evaluations as definitive

Are you saying that if my conditional type `A extends B ? X : Y` evaluates `Y`, I cannot rule out the case that `A extends  B`?

> Restrictive instantiations can and do occur

Can you give an example of the most restrictive and permissive instantiations for a generic type parameter? I have seen this terminology in the codebase, but I don't fully understand it.

Also, when does a restrictive instantiation occur?

> for *most* conditional types, they should be of the form `A extends B ? X : X | Y` to produce correct results.

Is this saying that most conditional types should be written in the form `A extends B ? X : (X | Y)` (parens added for clarity), meaning instead of something like
```ts
type Config<T extends "dev" | "stg" | "prod"> = T extends "prod" ? "safe config" : "allow experiments"
```
we should actually write
```ts
type Config<T extends "dev" | "stg" | "prod"> = T extends "prod" ? "safe config" : ("allow experiments" | "safe config")
```
otherwise `Config<"prod">` could still be judgmentally equal to `"allow experiments"`?
